api/src/main/java/org/apache/iceberg/SnapshotRef.java

```diff
@@ -0,0 +1,177 @@
+    return maxSnapshotAgeMs;
+  }
+
+  public Long maxRefAgeMs() {
+    return maxRefAgeMs;
+  }
+
+  public static Builder tagBuilder(long snapshotId) {
+    return builderFor(snapshotId, SnapshotRefType.TAG);
+  }
+
+  public static Builder branchBuilder(long snapshotId) {
+    return builderFor(snapshotId, SnapshotRefType.BRANCH);
+  }
+
+  public static Builder builderFrom(SnapshotRef ref) {
+    return new Builder(ref.type(), ref.snapshotId())
+        .minSnapshotsToKeep(ref.minSnapshotsToKeep())
+        .maxSnapshotAgeMs(ref.maxSnapshotAgeMs())
+        .maxRefAgeMs(ref.maxRefAgeMs());
+  }
+
+  /**
+   * Creates a ref builder from the given ref and its properties but the ref will now point to the given snapshotId.
+   *
+   * @param ref Ref to build from
+   * @param snapshotId snapshotID to use.
+   * @return ref builder with the same retention properties as given ref, but the ref will point to the passed in id
+   */
+  public static Builder builderFrom(SnapshotRef ref, long snapshotId) {
+    return new Builder(ref.type(), snapshotId)
+        .minSnapshotsToKeep(ref.minSnapshotsToKeep())
+        .maxSnapshotAgeMs(ref.maxSnapshotAgeMs())
+        .maxRefAgeMs(ref.maxRefAgeMs());
+  }
+
+  public static Builder builderFor(long snapshotId, SnapshotRefType type) {
+    return new Builder(type, snapshotId);
+  }
+
+  public static class Builder {
+
+    private final SnapshotRefType type;
+    private final long snapshotId;
+    private Integer minSnapshotsToKeep;
+    private Long maxSnapshotAgeMs;
+    private Long maxRefAgeMs;
+
+    Builder(SnapshotRefType type, long snapshotId) {
+      Preconditions.checkArgument(type != null, "Snapshot reference type must not be null");
+      this.type = type;
+      this.snapshotId = snapshotId;
+    }
+
+    public Builder minSnapshotsToKeep(Integer value) {
+      Preconditions.checkArgument(value == null || !type.equals(SnapshotRefType.TAG),
+          "Tags do not support setting minSnapshotsToKeep");
+      Preconditions.checkArgument(value == null || value > 0,
+          "Min snapshots to keep must be greater than 0");
+      this.minSnapshotsToKeep = value;
+      return this;
+    }
+
+    public Builder maxSnapshotAgeMs(Long value) {
+      Preconditions.checkArgument(value == null || !type.equals(SnapshotRefType.TAG),
+            "Tags do not support setting maxSnapshotAgeMs");
+      Preconditions.checkArgument(value == null || value > 0,
+          "Max snapshot age must be greater than 0");
+      this.maxSnapshotAgeMs = value;
+      return this;
+    }
+
+    public Builder maxRefAgeMs(Long value) {
+      Preconditions.checkArgument(value == null || value > 0, "Max reference age must be greater than 0");
+      this.maxRefAgeMs = value;
+      return this;
+    }
+
+    public SnapshotRef build() {
+      return new SnapshotRef(snapshotId, type, minSnapshotsToKeep, maxSnapshotAgeMs, maxRefAgeMs);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
Since the tests all check the individual values, `equals` and `hashCode` implementations aren't really needed. Are you planning on using them later?

The trouble with doing this right now is that you have to decide on what equality means for a `SnapshotRef`. This defaults to checking all fields, but it could be that it makes sense to use the name and the snapshot at some point in the future because the other 3 fields are metadata.

Normally, I recommend not making this decision until we need to. If you want to keep this implementation, I'm okay with that. It's up to you.